### PR TITLE
REGR: revert "CLN: _consolidate_inplace less" / fix regression in fillna()

### LINKED ISSUE #36495: REGR: fillna not filling NaNs after pivot without explicitly listing pivot values
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

Hi,

I have this (simplified) piece of code which seems to have stopped working correctly in Pandas 1.1:

```python
import pandas as pd

data = [
    [1, 1, 1, 1.],
    [2, 2, 2, 2.],
    [3, 3, 3, 3.],
]

df = pd.DataFrame(data, columns=['i1', 'i2', 'i3', 'f1'])
df = df.pivot('i1', 'i2')
df = df.fillna(0)
print(df)

```

#### Problem description

This is the output in Pandas 1.1.2:
```
     i3             f1
i2    1    2    3    1    2    3
i1
1   1.0  NaN  NaN  1.0  0.0  0.0
2   NaN  2.0  NaN  0.0  2.0  0.0
3   NaN  NaN  3.0  0.0  0.0  3.0
```

It contains NaNs after using fillna. It can be fixed by explicitly providing remaining columns as values to pivot (which should be the default):
`df = df.pivot('i1', 'i2', values=['i3', 'f1'])`

It also works if all columns are of the same type, like when I change f1 to contain ints instead of floats. In that case I don't need to specify the values.

#### Expected Output

This is the output on Pandas 1.0.5 and 0.25.3:
```
     i3             f1
i2    1    2    3    1    2    3
i1
1   1.0  0.0  0.0  1.0  0.0  0.0
2   0.0  2.0  0.0  0.0  2.0  0.0
3   0.0  0.0  3.0  0.0  0.0  3.0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2a7d3326dee660824a8433ffd01065f8ac37f7d6
python           : 3.7.8.final.0
python-bits      : 32
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : None.None

pandas           : 1.1.2
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 50.3.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : 1.3.18
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None
None

</details>

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Reverts pandas-dev/pandas#34389

Closes #36495

### CODE REVIEW

**Comment by jbrockmendel** [comment]
do the asvs show any of the perf issues mentioned in #34389?

**Comment by jorisvandenbossche** [comment]
The benchmark server isn't running at the moment, see https://github.com/pandas-dev/pandas/pull/34389#issuecomment-634654988

**Comment by jbrockmendel** [comment]
i was hoping you'd be able to run them locally.  (as mentioned in the other thread, im wrestling with hardware issues ATM)

**Comment by jorisvandenbossche** [comment]
Sorry, I currently don't have time to run the full ASV suite.

**Comment by jbrockmendel** [comment]
ive gotten asv working locally, will do a run on #34389 if were OK on holding off on merging this

**Comment by jorisvandenbossche** [comment]
As I also said in #34389, running asv is not necessarily sufficient, as it not sure that the functions that are tested are actually covered (it's not that our benchmark suite is 100% covering all cases). 
I think you need to specifically check the functions you changed and compare them with and without consolidation.

**Comment by jorisvandenbossche** [comment]
Or, for some of them it might be sufficient to simply check in the code if it would matter or not. 
For example, I was looking at the `take` example, and `NDFrame.take` calls `BlockManager.take`, and the first thing that does is also consolidate inplace. 
So in such as case, of course the `consolidate_inplace` in `NDFrame.take` can be removed as a code clean-up, and it also won't matter anything performance wise.

**Comment by jbrockmendel** [comment]
> So in such as case, of course the consolidate_inplace in NDFrame.take can be removed as a code clean-up, and it also won't matter anything performance wise.

Let's un-revert that here and whittle down the places that merit double-checking

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -667,6 +675,8 @@ def is_mixed_type(self) -> bool:
 
     @property
     def is_numeric_mixed_type(self) -> bool:
+        # Warning, consolidation needs to get checked upstairs
+        self._consolidate_inplace()
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:679]: this one should be pretty obviously innocuous

**Comment by jbrockmendel** [comment]
I'm not seeing any consistent pattern in asv results

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/managers.py]: re-reverted

**Comment by jorisvandenbossche** [comment]
> Let's un-revert that here and whittle down the places that merit double-checking

Done

**Comment by jorisvandenbossche** [comment]
@jbrockmendel what's the status here? (I re-reverted the ones discussed above)

Do we continue with this PR reverting those, or do you have time to check those cases?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche is this still active?

**Comment by simonjayhawkins** [comment]
I milestoned this 1.1.1 since it reverted a PR that caused a regression #35488. #35578 has been merged (and backported) to fix #35488 so am removing milestone.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche is this still active/needed?

**Comment by jorisvandenbossche** [comment]
There are still open regression reports related to this, eg https://github.com/pandas-dev/pandas/issues/35521

File: pandas/core/generic.py [hunk 2]
```diff
@@ -5428,6 +5430,7 @@ def values(self) -> np.ndarray:
                ['lion', 80.5, 1],
                ['monkey', nan, None]], dtype=object)
         """
+        self._consolidate_inplace()
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:5433]: this one has already been reverted (#34999)

**Comment by simonjayhawkins** [comment]
@jorisvandenbossche do you want to push for this in 1.1.2

**Comment by simonjayhawkins** [comment]
or can I move to 1.1.3

**Comment by jorisvandenbossche** [comment]
The revert for the issue mentioned above was merged in a separate PR, so moving to 1.1.3 (would need to take the time to investigate the others)

**Comment by simonjayhawkins** [comment]
another issue surfaced #36495

**Comment by jreback** [comment]
if someone wants to merge master can consider this

**Comment by simonjayhawkins** [comment]
will add test and release note for #36495 here too

**Comment by simonjayhawkins** [comment]
```
=========================== short test summary info ============================
FAILED pandas/tests/groupby/test_groupby.py::test_groupby_duplicate_columns
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_agg_split_object_part_datetime
=== 2 failed, 72815 passed, 4137 skipped, 1129 xfailed in 506.02s (0:08:26) ====
```

**Comment by simonjayhawkins** [comment]
it appears that #35839 and #35854 have been built on top of the changes that are being reverted.

**Comment by simonjayhawkins** [comment]
moving to 1.1.4

**Comment by jorisvandenbossche** [comment]
Do we still want to revert the remaining others in here as well? (except the one in apply, which causes the failures)

**Comment by simonjayhawkins** [comment]
> Do we still want to revert the remaining others in here as well? (except the one in apply, which causes the failures)

good point. I think only the one is causing the test fails

**Comment by jbrockmendel** [comment]
This doesn't add any tests.  What is this fixing?

**Comment by simonjayhawkins** [comment]
not seen this one. restarting

```
=========================== short test summary info ===========================
FAILED pandas/tests/io/test_sql.py::TestSQLApi::test_warning_case_insensitive_table_name[load_iris_data0]
= 1 failed, 75531 passed, 1816 skipped, 1165 xfailed, 58 warnings in 1103.51s (0:18:23) =
```

**Comment by jorisvandenbossche** [comment]
> This doesn't add any tests. What is this fixing?

Well, the PR it is reverting also didn't add any tests.

But so, the reason I asked whether we still want to go forward with this PR: 1) my original reason for proposing to revert the original PR still stands (i.e. the reason I opened the PR, eg https://github.com/pandas-dev/pandas/pull/34407#issuecomment-634881185, about the need to check the performance impact of those changes), and 2) removing the consolidation in some places caused several regressions. Multiple of the cases where consolidation was removed in the original PR already have been reverted in the mean time. 
Given the apparently risky change, and given that there was not (IMO) a clear motivation for merging the original PR (no good evidence it's actually faster/better), it might be "safer" to also revert the remaining cases to avoid more regressions.  

This last sentence if of course purely hypothetical (maybe all regression related to the original PR now have been fixed), so therefore it was also a *question* whether we still want to go forward with this PR. I am not fully sure myself if that is a good idea, certainly given that the original PR has already been quite some time in master now, while if we revert it, it would only be a very short time in master before 1.1.3

**Comment by simonjayhawkins** [comment]
looks like needs more discussion. moving to 1.1.4

**Comment by jreback** [comment]
i am not sure it makes sense to push this w/o a clear patch backed by tests. As i am not sure what excatly this is fixing.

**Comment by jreback** [comment]
moving this off 1.1.4 as not really clear metrics on how to evaluate this.

**Comment by jreback** [comment]
not sure what the point of the PR Is any longer.

**Comment by jorisvandenbossche** [comment]
> not sure what the point of the PR Is any longer.

There are still open regressions related to this (#36495). 
(will take a look again in a few days)

**Comment by jreback** [comment]
> > not sure what the point of the PR Is any longer.
> 
> There are still open regressions related to this (#36495).
> (will take a look again in a few days)

maybe so but we still don't have any testing. so -1 on including this on 1.1.5 at this point.

**Comment by jorisvandenbossche** [comment]
> but we still don't have any testing

I added a test for the regression reported in https://github.com/pandas-dev/pandas/issues/36495, which is fixed by this

File: pandas/tests/frame/methods/test_fillna.py [hunk 3]
```diff
@@ -524,3 +524,18 @@ def test_fill_corner(self, float_frame, float_string_frame):
...
+        [
+            [1, 1, 1, 1.0],
+            [2, 2, 2, 2.0],
+            [3, 3, 3, 3.0],
+        ],
+        columns=["i1", "i2", "i3", "f1"],
+    )
+    df_nonconsol = df.pivot("i1", "i2")
+    result = df_nonconsol.fillna(0)
+    assert result.isna().sum().sum() == 0
```
Comment (simonjayhawkins) [inline pandas/tests/frame/methods/test_fillna.py:541]: there's a test in #36668 if you wanted a more explicit assertion

**Comment by jbrockmendel** [comment]
> I added a test for the regression reported in #36495, which is fixed by this

great.  there are four consolidations this adds.  is any one in particular responsible for fixing this bug?

**Comment by jorisvandenbossche** [comment]
I suppose it is the consolidation in `fillna` that specifically fixes the reported regression. 

I could do a PR with specifically only that change, but as mentioned above (https://github.com/pandas-dev/pandas/pull/34407#issuecomment-699541453), I think it can't hurt to change the other cases as well. Since the original PR #34389, we  now already reverted parts of that change in 3 other PRs that were fixing regressions, and this is a 4th.

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Comment by jbrockmendel** [comment]
AFAICT the consolidation in quantile is never reached.  am i missing something?

**Comment by jorisvandenbossche** [comment]
The default of the keyword is True, so it should be reached? (since we never specify this keyword anywhere at the moment)

**Pull request merged**